pubsub/tests/unit/subscriber_test.py

```diff
@@ -0,0 +1,424 @@
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_cache_outdated_false(subscriber_client):
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 0)
+        cache.last_refresh = time.perf_counter()
+        assert cache.cache_outdated()
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_cache_outdated_true(subscriber_client):
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 0)
+        cache.ack_deadline = time.perf_counter()
+        assert cache.cache_outdated()
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_refresh_updates_value_and_last_refresh(
+        subscriber_client
+    ):
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 1)
+        await cache.refresh()
+        assert cache.ack_deadline == 42
+        assert cache.last_refresh
+        subscriber_client.get_subscription.assert_called_once_with(
+            'fake_subscription')
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_refresh_is_cool_about_failures(
+        subscriber_client
+    ):
+        subscriber_client.get_subscription = MagicMock(
+            side_effect=RuntimeError)
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 1)
+        cache.ack_deadline = 55.0
+        await cache.refresh()
+        assert cache.ack_deadline == 55.0
+        assert cache.last_refresh
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_get_calls_refresh_first_time(
+        subscriber_client
+    ):
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 1)
+        assert await cache.get() == 42
+        assert cache.last_refresh
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_get_no_call_if_not_outdated(
+        subscriber_client
+    ):
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 1000)
+        cache.ack_deadline = 33
+        cache.last_refresh = time.perf_counter()
+        assert await cache.get() == 33
+        subscriber_client.get_subscription.assert_not_called()
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_get_refreshes_if_outdated(
+            subscriber_client):
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 0)
+        cache.ack_deadline = 33
+        assert await cache.get() == 42
+        assert cache.last_refresh
+        subscriber_client.get_subscription.assert_called_once()
+
+    @pytest.mark.asyncio
+    async def test_ack_deadline_cache_first_get_failed(subscriber_client):
+        subscriber_client.get_subscription = MagicMock(
+            side_effect=RuntimeError)
+        cache = AckDeadlineCache(
+            subscriber_client, 'fake_subscription', 10)
+        assert await cache.get() == float('inf')
+        assert cache.last_refresh
```

Comment:
Okay I think this is fine because I didn't yet take a look at `cache.get()`, where a refresh will be tried as long as the ack deadline keeps coming up as "not having been set" (so you can ignore my previous comment):

```python
        async def get(self) -> float:
            if self.ack_deadline == float('inf') or self.cache_outdated():
                await self.refresh()
            return self.ack_deadline
```

BTW can the `ack_deadline` ever be legitimately `float('inf')`?

Reply:
you're right, we refresh cache every time to limit our potential network output. There is actually a bug here. For the very first fetch we will retry every time thanks to `if self.ack_deadline == float('inf') or self.cache_outdated()` condition. I had a pleasure to see an impact today on staging -> application very quickly becomes IO bound.

>BTW can the ack_deadline ever be legitimately float('inf')?

nah, max value is 600s